### PR TITLE
Nullable annotate System.Security.Principal.Windows

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Contributes to https://github.com/dotnet/runtime/issues/2339

### CODE REVIEW

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs [hunk 1]
```diff
@@ -320,19 +320,19 @@ void IDeserializationCallback.OnDeserialization(object sender)
...
 
-        public static WindowsIdentity GetCurrent(bool ifImpersonating)
+        public static WindowsIdentity? GetCurrent(bool ifImpersonating)
         {
             return GetCurrentInternal(TokenAccessLevels.MaximumAllowed, ifImpersonating);
         }
 
 
-        public static WindowsIdentity GetCurrent(TokenAccessLevels desiredAccess)
+        public static WindowsIdentity? GetCurrent(TokenAccessLevels desiredAccess)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: These three methods signatures are the only ones that have impact in the public API of this library. It is not clear to me whether I should annotate as non-nullable, however it is evident from WindowsPrincipal.cs#L179 that the parameterless method is expected to be non-nullable. I'm not familiar with this component, so not sure what the expected invariant should be here.

Comment (jkotas) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: Notice that `GetCurrentInternal(TokenAccessLevels desiredAccess, bool threadOnly)` can return null only when `threadOnly == true`. This should give you decoder ring which variants of `GetCurrent` can may return null and when.

File: src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs [hunk 2]
```diff
@@ -20,19 +20,19 @@ namespace System.Security.Principal
...
     {
         internal IdentityReference() { }
         public abstract string Value { get; }
-        public abstract override bool Equals(object o);
+        public abstract override bool Equals(object? o);
         public abstract override int GetHashCode();
         public abstract bool IsValidTargetType(System.Type targetType);
-        public static bool operator ==(System.Security.Principal.IdentityReference left, System.Security.Principal.IdentityReference right) { throw null; }
+        public static bool operator ==(System.Security.Principal.IdentityReference? left, System.Security.Principal.IdentityReference? right) { throw null; }
         public static bool operator !=(System.Security.Principal.IdentityReference left, System.Security.Principal.IdentityReference right) { throw null; }
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: Shouldn't != also have nullable arguments?

File: src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs [hunk 3]
```diff
@@ -59,10 +59,10 @@ public sealed partial class NTAccount : System.Security.Principal.IdentityRefere
         public NTAccount(string name) { }
         public NTAccount(string domainName, string accountName) { }
         public override string Value { get { throw null; } }
-        public override bool Equals(object o) { throw null; }
+        public override bool Equals(object? o) { throw null; }
         public override int GetHashCode() { throw null; }
         public override bool IsValidTargetType(System.Type targetType) { throw null; }
-        public static bool operator ==(System.Security.Principal.NTAccount left, System.Security.Principal.NTAccount right) { throw null; }
+        public static bool operator ==(System.Security.Principal.NTAccount? left, System.Security.Principal.NTAccount? right) { throw null; }
         public static bool operator !=(System.Security.Principal.NTAccount left, System.Security.Principal.NTAccount right) { throw null; }
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: Ditto

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: Right, so makes sense to `GetCurrent(bool)` as the only method with nullable return?

File: src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs [hunk 4]
```diff
@@ -171,8 +171,6 @@ public enum WellKnownSidType
         BuiltinPerformanceMonitoringUsersSid = 57,
         BuiltinPerformanceLoggingUsersSid = 58,
         BuiltinAuthorizationAccessSid = 59,
-        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
-        [System.ObsoleteAttribute("This member has been depcreated and is only maintained for backwards compatability. WellKnownSidType values greater than MaxDefined may be defined in future releases.")]
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: Why are these being removed?

Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: depcreated => deprecated

File: src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs [hunk 5]
```diff
@@ -210,7 +208,6 @@ public enum WellKnownSidType
         WinCapabilityEnterpriseAuthenticationSid = 93,
         WinCapabilityRemovableStorageSid = 94,
     }
-    [System.Runtime.InteropServices.ComVisibleAttribute(true)]
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: Have we removed ComVisibleAttributes from elsewhere?  I can't remember now.

File: src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs [hunk 6]
```diff
@@ -241,31 +238,31 @@ public partial class WindowsIdentity : System.Security.Claims.ClaimsIdentity, Sy
...
+        public System.Security.Principal.SecurityIdentifier? User { get { throw null; } }
         public virtual System.Collections.Generic.IEnumerable<System.Security.Claims.Claim> UserClaims { get { throw null; } }
         public override System.Security.Claims.ClaimsIdentity Clone() { throw null; }
         public void Dispose() { }
         protected virtual void Dispose(bool disposing) { }
         public static System.Security.Principal.WindowsIdentity GetAnonymous() { throw null; }
-        public static System.Security.Principal.WindowsIdentity GetCurrent() { throw null; }
-        public static System.Security.Principal.WindowsIdentity GetCurrent(bool ifImpersonating) { throw null; }
-        public static System.Security.Principal.WindowsIdentity GetCurrent(System.Security.Principal.TokenAccessLevels desiredAccess) { throw null; }
+        public static System.Security.Principal.WindowsIdentity? GetCurrent() { throw null; }
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: This won't return null.

File: src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: It should, thanks

File: src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs [hunk 7]
```diff
@@ -241,31 +238,31 @@ public partial class WindowsIdentity : System.Security.Claims.ClaimsIdentity, Sy
...
+        public static System.Security.Principal.WindowsIdentity? GetCurrent(bool ifImpersonating) { throw null; }
+        public static System.Security.Principal.WindowsIdentity? GetCurrent(System.Security.Principal.TokenAccessLevels desiredAccess) { throw null; }
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: This won't return null.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/NTAccount.cs [hunk 8]
```diff
@@ -208,7 +208,7 @@ internal static IdentityReferenceCollection Translate(IdentityReferenceCollectio
             }
             else
             {
-                return (left.ToString().Equals(right.ToString(), StringComparison.OrdinalIgnoreCase));
+                return (left!.ToString().Equals(right!.ToString(), StringComparison.OrdinalIgnoreCase));
             }
         }
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/NTAccount.cs]: All of the != operators should have the same nullability as the == operators.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs [hunk 9]
```diff
@@ -288,16 +288,18 @@ public sealed class SecurityIdentifier : IdentityReference, IComparable<Security
         //
 
         private IdentifierAuthority _identifierAuthority;
-        private int[] _subAuthorities;
-        private byte[] _binaryForm;
-        private SecurityIdentifier _accountDomainSid;
+        // values guaranteed to be non-null on account of CreateFromParts()
+        // method called by every constructor in this class.
+        private int[] _subAuthorities = null! ;
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs]:
```suggestion
        private int[] _subAuthorities = null!;
```

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs [hunk 10]
```diff
@@ -288,16 +288,18 @@ public sealed class SecurityIdentifier : IdentityReference, IComparable<Security
...
+        private byte[] _binaryForm = null! ;
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs]:
```suggestion
        private byte[] _binaryForm = null!;
```

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: Correct.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs [hunk 11]
```diff
@@ -574,15 +574,17 @@ internal string GetName()
                 // revert thread impersonation for the duration of the call to get the name.
                 RunImpersonated(SafeAccessTokenHandle.InvalidHandle, delegate
                 {
-                    NTAccount ntAccount = User.Translate(typeof(NTAccount)) as NTAccount;
-                    _name = ntAccount.ToString();
+                    NTAccount? ntAccount = User!.Translate(typeof(NTAccount)) as NTAccount;
+                    _name = ntAccount!.ToString();
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: This is a strange construction.  Doesn't seem worth changing it to improve the code, but, yeah.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs [hunk 12]
```diff
@@ -592,17 +594,17 @@ public SecurityIdentifier Owner
 
                 if (_owner == null)
                 {
-                    using (SafeLocalAllocHandle tokenOwner = GetTokenInformation(_safeTokenHandle, TokenInformationClass.TokenOwner))
+                    using (SafeLocalAllocHandle? tokenOwner = GetTokenInformation(_safeTokenHandle, TokenInformationClass.TokenOwner))
                     {
-                        _owner = new SecurityIdentifier(tokenOwner.Read<IntPtr>(0));
+                        _owner = new SecurityIdentifier(tokenOwner!.Read<IntPtr>(0));
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: The call to GetTokenInformation won't return null because true isn't being passed to the last optional argument.  It'd be better to just `!` on the GetTokenInformation call, and remove the `?` on `SafeLocalAllocHandle` and the `!` on `tokenOwner`.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs [hunk 13]
```diff
@@ -861,9 +863,9 @@ private static SafeAccessTokenHandle GetCurrentProcessToken(TokenAccessLevels de
         {
             Debug.Assert(!_safeTokenHandle.IsInvalid && !_safeTokenHandle.IsClosed, "!m_safeTokenHandle.IsInvalid && !m_safeTokenHandle.IsClosed");
 
-            using (SafeLocalAllocHandle information = GetTokenInformation(_safeTokenHandle, tokenInformationClass))
+            using (SafeLocalAllocHandle? information = GetTokenInformation(_safeTokenHandle, tokenInformationClass))
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: Same as earlier, for all such cases.

File: src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs (hunk 4 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: Looks like reference source generation snafu. Added back manually.

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: Fixed

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsPrincipal.cs [hunk 14]
```diff
@@ -24,7 +24,7 @@ public enum WindowsBuiltInRole
 
     public class WindowsPrincipal : ClaimsPrincipal
     {
-        private readonly WindowsIdentity _identity = null;
+        private readonly WindowsIdentity _identity = null!;
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsPrincipal.cs]: Where is the `private WindowsPrincipal() { }` constructor used?  That seems to be the only reason this `= null!` is needed, and I don't see anyone using it (plus the type has other ctors that will prevent a default public one from being added).

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsPrincipal.cs [hunk 15]
```diff
@@ -176,6 +174,6 @@ public virtual bool IsInRole(SecurityIdentifier sid)
         }
 
         // This is called by AppDomain.GetThreadPrincipal() via reflection.
-        private static IPrincipal GetDefaultInstance() => new WindowsPrincipal(WindowsIdentity.GetCurrent());
+        private static IPrincipal GetDefaultInstance() => new WindowsPrincipal(WindowsIdentity.GetCurrent()!);
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsPrincipal.cs]: This `!` shouldn't be necessary.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsPrincipal.cs (hunk 14 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsPrincipal.cs]: It's dead code, but I opted for that approach instead of removing members.

Comment (stephentoub) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsPrincipal.cs]: Let's remove the dead code.  Thanks.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs (hunk 12 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: Good call. I think I might explicitly pass that optional argument for better emphasis.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsPrincipal.cs (hunk 14 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsPrincipal.cs]: Done.

**Comment by eiriktsarpalis** [comment]
Looks like `System.Security.AccessControl` has a dependency on this library, which is causing some of the build failures.

File: src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs (hunk 5 above)
Comment (jkotas) [reply src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: Yes, we have deleted it from (close to) everywhere. This one looks like oversight. It should be fine to delete it both from here and from the implementation.

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: Done

File: src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs [hunk 16]
```diff
@@ -171,8 +171,6 @@ public enum WellKnownSidType
         BuiltinPerformanceMonitoringUsersSid = 57,
         BuiltinPerformanceLoggingUsersSid = 58,
         BuiltinAuthorizationAccessSid = 59,
-        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
```
Comment (jkotas) [inline src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: Looks like this got removed again unintentionally.

File: src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs [hunk 17]
```diff
@@ -241,31 +238,31 @@ public partial class WindowsIdentity : System.Security.Claims.ClaimsIdentity, Sy
...
+        public static System.Threading.Tasks.Task RunImpersonatedAsync(Microsoft.Win32.SafeHandles.SafeAccessTokenHandle safeAccessTokenHandle, System.Func<System.Threading.Tasks.Task> func) { throw null; }
+        public static System.Threading.Tasks.Task<T> RunImpersonatedAsync<T>(Microsoft.Win32.SafeHandles.SafeAccessTokenHandle safeAccessTokenHandle, System.Func<System.Threading.Tasks.Task<T>> func) { throw null; }
         public static T RunImpersonated<T>(Microsoft.Win32.SafeHandles.SafeAccessTokenHandle safeAccessTokenHandle, System.Func<T> func) { throw null; }
-        public static System.Threading.Tasks.Task RunImpersonatedAsync(Microsoft.Win32.SafeHandles.SafeAccessTokenHandle safeAccessTokenHandle, Func<System.Threading.Tasks.Task> func) { throw null; }
-        public static System.Threading.Tasks.Task<T> RunImpersonatedAsync<T>(Microsoft.Win32.SafeHandles.SafeAccessTokenHandle safeAccessTokenHandle, Func<System.Threading.Tasks.Task<T>> func) { throw null; }
         void System.Runtime.Serialization.IDeserializationCallback.OnDeserialization(object sender) { }
```
Comment (eiriktsarpalis) [inline src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: Hmm, looks like the source generator is omitting the nullable annotation for the `sender` argument. Compare with https://github.com/dotnet/runtime/pull/2347/files#diff-7c0b356a7bada939e2b19732c1c3334dR314 @safern any thoughts?

Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs]: Seems that setting it manually doesn't fix the issue on non-windows machines, on account of auto-generated `*.notsupported.cs` files: https://dev.azure.com/dnceng/public/_build/results?buildId=501665&view=logs&j=0ce95488-260d-5e86-47b9-82bc4e703e0b&t=fc172bf4-4375-570c-6063-be42dda01a5a&l=816

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs [hunk 18]
```diff
@@ -288,16 +288,18 @@ public sealed class SecurityIdentifier : IdentityReference, IComparable<Security
         //
 
         private IdentifierAuthority _identifierAuthority;
-        private int[] _subAuthorities;
-        private byte[] _binaryForm;
-        private SecurityIdentifier _accountDomainSid;
+        // values guaranteed to be non-null on account of CreateFromParts()
+        // method called by every constructor in this class.
+        private int[] _subAuthorities = null!;
+        private byte[] _binaryForm = null!;
```
Comment (Wraith2) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs]: I'm not quite sure what this means, the comment indicates they're non-null typed but then it's changed change explicitly applies null to them? Why are the nulls needed?

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs [hunk 19]
```diff
@@ -394,7 +396,7 @@ private void CreateFromParts(IdentifierAuthority identifierAuthority, ReadOnlySp
             }
         }
 
-        private void CreateFromBinaryForm(byte[] binaryForm, int offset)
+        private void CreateFromBinaryForm(byte[]? binaryForm, int offset)
```
Comment (Wraith2) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs]: The `binaryForm` parameter is immediately checked for null and throws if that is the case so this method will never accept a null value. Doesn't that mean that the type should be non-null? I realise that a nullable array is needed later but another variable could be used for that.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs (hunk 18 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs]: Because the values are initialized by a method and not the constructor directly, the analyzer will complain about nullability unless overriden like above.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs (hunk 19 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs]: You're right that the signature of the method should be updated to reflect that check.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs (hunk 18 above)
Comment (Wraith2) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs]: So we'd need a way to express to the analyzer that the called method guarnetees that these non-argument members are non-null as a postcondition, which isn't currently possible. Or we could change create parts to take the assignment to the caller not the calle so it can see the assignment, i don't like the idea of rearranging code just to work around an analyzers limitation though.

Whether this is a (really tiny) problem depends on which is more important, cleaner code without null assignments or getting a clean slate on the analyzer.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs [hunk 20]
```diff
@@ -322,19 +322,22 @@ void IDeserializationCallback.OnDeserialization(object sender)
...
+
             return GetCurrentInternal(TokenAccessLevels.MaximumAllowed, ifImpersonating);
         }
 
 
         public static WindowsIdentity GetCurrent(TokenAccessLevels desiredAccess)
         {
-            return GetCurrentInternal(desiredAccess, false);
+            // not null when threadOnly argument is false
+            return GetCurrentInternal(desiredAccess, false)!;
```
Comment (Wraith2) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: It feels like there should be a way to express that "return is non-null when argument is true/false" but i see there isn't.

**Review by Wraith2 — COMMENTED** [review]
LGTM,

**Review by Wraith2 — APPROVED** [review]

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs (hunk 20 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: Yeah, something like that would be helpful here.

Comment (stephentoub) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: It is one of the attributes we considered and punted on initially (not just for true/false, but for any value that can be determined and compared statically). It shows up in other places, like in reflection methods that take a throw-if-not-found argument. That said, it hasn't come up frequently enough to warrant inclusion yet.
cc: @jcouv, @madstorgersen

**Comment by eiriktsarpalis** [comment]
Builds working now, this should be ready to go in.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs (hunk 19 above)
Comment (Wraith2) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/SID.cs]: Did you change this? I can't see it but it might be me know knowing how to use GitHub.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/NTAccount.cs [hunk 21]
```diff
@@ -137,9 +137,9 @@ public override IdentityReference Translate(Type targetType)
...
 
-        public override bool Equals(object o)
+        public override bool Equals(object? o)
         {
-            return (this == o as NTAccount); // invokes operator==
+            return o is NTAccount acc && this == acc; // invokes operator==
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/NTAccount.cs]: Why is this change needed?

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs [hunk 22]
```diff
@@ -36,28 +36,28 @@ public enum WindowsAccountType
...
-        private SecurityIdentifier _user = null;
-        private IdentityReferenceCollection _groups = null;
+        private string? _name = null;
+        private SecurityIdentifier? _owner = null;
+        private SecurityIdentifier? _user = null;
+        private IdentityReferenceCollection? _groups = null;
 
         private SafeAccessTokenHandle _safeTokenHandle = SafeAccessTokenHandle.InvalidHandle;
-        private readonly string _authType = null;
+        private readonly string? _authType = null;
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: We can delete all of these ` = null` initializations.

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs [hunk 23]
```diff
@@ -322,19 +322,22 @@ void IDeserializationCallback.OnDeserialization(object sender)
 
         public static WindowsIdentity GetCurrent()
         {
-            return GetCurrentInternal(TokenAccessLevels.MaximumAllowed, false);
+            // not null when threadOnly argument is false
+            return GetCurrentInternal(TokenAccessLevels.MaximumAllowed, false)!;
```
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: Nit:
```suggestion
            return GetCurrentInternal(TokenAccessLevels.MaximumAllowed, threadOnly:false)!;
```

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs (hunk 20 above)
Comment (stephentoub) [inline src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]:
```suggestion
            return GetCurrentInternal(desiredAccess, threadOnly: false)!;
```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/NTAccount.cs (hunk 21 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/NTAccount.cs]: It isn't. Reverting.

**Comment by eiriktsarpalis** [comment]
CI failures seem unrelated to change, merging away.
